src/Client/ClientBase.cpp

```diff
@@ -0,0 +1,1530 @@
+        auto in_buffer = wrapReadBufferWithCompressionMethod(std::make_unique<ReadBufferFromFile>(in_file), chooseCompressionMethod(in_file, compression_method));
+
+        try
+        {
+            sendDataFrom(*in_buffer, sample, columns_description, parsed_query);
+        }
+        catch (Exception & e)
+        {
+            e.addMessage("data for INSERT was parsed from file");
+            throw;
+        }
+    }
+    else if (parsed_insert_query->data)
+    {
+        /// Send data contained in the query.
+        ReadBufferFromMemory data_in(parsed_insert_query->data, parsed_insert_query->end - parsed_insert_query->data);
+        try
+        {
+            sendDataFrom(data_in, sample, columns_description, parsed_query);
+        }
+        catch (Exception & e)
+        {
+            /// The following query will use data from input
+            //      "INSERT INTO data FORMAT TSV\n " < data.csv
+            //  And may be pretty hard to debug, so add information about data source to make it easier.
+            e.addMessage("data for INSERT was parsed from query");
+            throw;
+        }
+        // Remember where the data ended. We use this info later to determine
+        // where the next query begins.
+        parsed_insert_query->end = parsed_insert_query->data + data_in.count();
+    }
+    else if (!is_interactive)
+    {
+        if (need_render_progress)
+        {
+            /// Add callback to track reading from fd.
+            std_in.setProgressCallback(global_context);
+        }
+
+        /// Send data read from stdin.
+        try
+        {
+            sendDataFrom(std_in, sample, columns_description, parsed_query);
+        }
+        catch (Exception & e)
+        {
+            e.addMessage("data for INSERT was parsed from stdin");
+            throw;
+        }
+    }
+    else
+        throw Exception("No data to insert", ErrorCodes::NO_DATA_TO_INSERT);
+}
+
+
+void ClientBase::sendDataFrom(ReadBuffer & buf, Block & sample, const ColumnsDescription & columns_description, ASTPtr parsed_query)
+{
+    String current_format = insert_format;
+
+    /// Data format can be specified in the INSERT query.
+    if (const auto * insert = parsed_query->as<ASTInsertQuery>())
+    {
+        if (!insert->format.empty())
+            current_format = insert->format;
+    }
+
+    auto source = FormatFactory::instance().getInput(current_format, buf, sample, global_context, insert_format_max_block_size);
+    Pipe pipe(source);
+
+    if (columns_description.hasDefaults())
+    {
+        pipe.addSimpleTransform([&](const Block & header)
+        {
+            return std::make_shared<AddingDefaultsTransform>(header, columns_description, *source, global_context);
+        });
+    }
+
+    QueryPipeline pipeline;
+    pipeline.init(std::move(pipe));
+    PullingAsyncPipelineExecutor executor(pipeline);
+
+    Block block;
+    while (executor.pull(block))
```

Comment:
btw @azat I remember your recent change to add in this place `full_query` to exception message, but not sure how to put it nicely now, because `full_query` variable is a local variable now, I really do no want to make it a class field as it was before and as well It will take a very long path to forward `full_query` to this place. But on the other hand I once found your addition of `full_query` to exception message useful in tests, so I cannot decide now what to do...

Reply:
> but not sure how to put it nicely now

Maybe just add `try`/`catch` into `ClientBase::processInsertQuery`?

> I really do no want to make it a class field as it was before 

Indeed, I also think that query should not be a class member.

Reply:
> Maybe just add try/catch into ClientBase::processInsertQuery?

Then also `processOrdinaryQuery`, because will look strange to add only for the first.. Then then we can say why not just wrap all processSingleQuery. Probably will do this way, hope not too many tests will need to be updated.

Reply:
@tavplubix what do you think?